---
version: '3.3'

services:
  app:
    image: docker.elastic.co/logstash/logstash:${STACK_VERSION}
    networks:
      - freer_network
    command: 'logstash -f /usr/share/logstash/pipeline/logstash.conf'
    environment:
      - xpack.monitoring.enabled=true
      - xpack.monitoring.elasticsearch.hosts=https://${ELASTICSEARCH_HOST}:9200
      - xpack.monitoring.elasticsearch.ssl.certificate_authority=/run/secrets/elasticsearch_single_node_cert
      - xpack.monitoring.elasticsearch.username=${LOGSTASH_USER}
      - xpack.monitoring.elasticsearch.password=${LOGSTASH_PASSWORD}
      - xpack.monitoring.collection.interval=10s
      - xpack.monitoring.collection.pipeline.details.enabled=true
    ports:
      - 12201:12201/udp
    deploy:
      mode: replicated
      replicas: 2
      placement:
        max_replicas_per_node: 1
        constraints:
          - node.labels.elk == true
      restart_policy:
        condition: on-failure 
        max_attempts: 3 
      resources:
        limits:
          memory: 1gb
        reservations:
          memory: 1gb
    configs:
      - source: logstash.conf.v4
        target: /usr/share/logstash/pipeline/logstash.conf
    secrets:
      - source: elasticsearch_single_node_cert
        target: elasticsearch_single_node_cert
        # https://discuss.elastic.co/t/docker-setup-generated-certs-logstash-permission-denied/301133
        # https://www.elastic.co/guide/en/logstash/current/docker-config.html
        # https://docs.docker.com/compose/compose-file/compose-file-v3/#secrets
        uid: '1000'
        gid: '1000'
        
# have to reference external network, otherwise a new scoped network is
# created during stack deploy.
# https://docs.docker.com/compose/compose-file/#external
networks:
  freer_network:
    external: true

configs:
  logstash.conf.v4:
    external: true

secrets:
  elasticsearch_single_node_cert:
    external: true
